presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SimplifyExpressions.java

```diff
@@ -271,4 +278,75 @@ public Expression rewriteLogicalBinaryExpression(LogicalBinaryExpression node, N
                     .collect(toImmutableList());
         }
     }
+
+    private static class RemoveIdentityCastContext
+    {
+        private final Map<Symbol, Expression> expressionAssignments;
+        private final Map<Symbol, Type> typeAssignments;
+
+        public RemoveIdentityCastContext(Map<Symbol, Expression> expressionAssignments,
+                                         Map<Symbol, Type> typeAssignments)
+        {
+            this.expressionAssignments = expressionAssignments;
+            this.typeAssignments = typeAssignments;
+        }
+
+        public Map<Symbol, Expression> getExpressionAssignments()
+        {
+            return expressionAssignments;
+        }
+
+        public Map<Symbol, Type> getTypeAssignments()
+        {
+            return typeAssignments;
+        }
+    }
+
+    private static class RemoveIdentityCastsRewriter
+            extends ExpressionRewriter<RemoveIdentityCastContext>
+    {
+        @Override
+        public Expression rewriteExpression(Expression node, RemoveIdentityCastContext context,
+                                            ExpressionTreeRewriter<RemoveIdentityCastContext> treeRewriter)
+        {
+            if (context.getExpressionAssignments() == null || context.getTypeAssignments() == null) {
+                return node;
+            }
+            else {
+                Map<Symbol, Type> typeAssignments = context.getTypeAssignments();
+                for (Map.Entry<Symbol, Expression> expressionAssignment : context.getExpressionAssignments().entrySet()) {
```

Comment:
So the problem I have is that `simplifyExpressions` is automatically called on every `Expression` (value, not a key) in the `expressionAssignments` map. I can't pass context or the associated `Symbol` key for the map to `simplifyExpressions`. 

So the function iterates through all of the assignments to find the `Symbol` key associated with the `Expression` value. The symbol is needed to resolve the type that the data will be casted to (as the type map maps from `Symbol` -> `Type`).

The `Cast` expression does have a `type` field, but it's stored as a `String` not a `Type` and so it doesn't behave as expected.